File: translated/f-to-c/DRB039-truedepsingleelement-orig-yes.c
/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.

SPDX-License-Identifier: (BSD-3-Clause)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

//Data race pair: a[i]@24:9:W vs. a[0]@24:16:R

#include <omp.h>
#include <stdio.h>
#include <stdlib.h>

int main()
{
    int len, i;
    int* a;

    len = 1000;
    a = (int*)malloc(len * sizeof(int));

    a[0] = 2;

    #pragma omp parallel for
    for (i = 0; i < len; i++) {
        a[i] = a[i] + a[0];
    }

    printf("a(500) = %d\n", a[499]);

    free(a);

    return 0;
} 

File: translated/f-to-go-automodel-bad/DRB039-truedepsingleelement-orig-yes.go
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
// and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.
//
// SPDX-License-Identifier: (BSD-3-Clause)
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

//Data race pair: a[i]@24:9:W vs. a[0]@24:16:R

package main

import (
	"fmt"
	"runtime"
	"sync"
)

func main() {
	var length int
	var a []int

	length = 1000
	a = make([]int, length)

	a[0] = 2

	//$omp parallel do
	var wg sync.WaitGroup
	numCPU := runtime.NumCPU()
	chunkSize := length / numCPU
	if chunkSize < 1 {
		chunkSize = 1
	}

	for start := 1; start <= length; start += chunkSize {
		end := start + chunkSize - 1
		if end > length {
			end = length
		}
		wg.Add(1)
		go func(start, end int) {
			defer wg.Done()
			for i := start; i <= end; i++ {
				a[i-1] = a[i-1] + a[0]
			}
		}(start, end)
	}
	wg.Wait()
	//$omp end parallel do

	fmt.Printf("a(500) =%3d\n", a[499])
}
